### PR TITLE
definitions for linux hardware timestamping

### PR DESCRIPTION
**Pull request by folkertdev**
Definitions can be found here https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/net_tstamp.h#L76

These definitions are relevant for (PTP and NTP) hardware timestamping

### CODE REVIEW

**Comment by folkertdev** [comment]
I'm not sure what this CI error means

```
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35676:80: error: 'HWTSTAMP_FLAG_BONDED_PHC_INDEX' undeclared here (not in a function)
  cargo:warning=35676 |             static const int __test_const_HWTSTAMP_FLAG_BONDED_PHC_INDEX_val = HWTSTAMP_FLAG_BONDED_PHC_INDEX;
  cargo:warning=      |                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35682:68: error: 'HWTSTAMP_FLAG_LAST' undeclared here (not in a function)
  cargo:warning=35682 |             static const int __test_const_HWTSTAMP_FLAG_LAST_val = HWTSTAMP_FLAG_LAST;
  cargo:warning=      |                                                                    ^~~~~~~~~~~~~~~~~~
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35688:68: error: 'HWTSTAMP_FLAG_MASK' undeclared here (not in a function)
  cargo:warning=35688 |             static const int __test_const_HWTSTAMP_FLAG_MASK_val = HWTSTAMP_FLAG_MASK;
  cargo:warning=      |                                                                    ^~~~~~~~~~~~~~~~~~
  cargo:warning=cc1: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
```

maybe the system libc on CI does not (yet?) have these symbols? They are not important to me so I'd be OK with removing them (just thought I might as well add them for completeness)

**Comment by JohnTitor** [comment]
> maybe the system libc on CI does not (yet?) have these symbols?

Correct.

Could you squash commits into one?

**Review by JohnTitor — APPROVED** [review]

**Comment by JohnTitor** [comment]
Thank you, @bors r+

**Comment by folkertdev** [comment]
I'm not sure what went wrong here. I assume the action needs to just be re-run?

**Comment by folkertdev** [comment]
I added two extra constants that are needed (`SIOCSHWTSTAMP` and `SIOCGHWTSTAMP`) and rebased. Maybe that makes bors happier?

though CI seems to be broken at the moment.

**Comment by JohnTitor** [comment]
The FreeBSD failure is unrelated and it's already fixed on master.
But https://github.com/rust-lang/libc/pull/3155#issuecomment-1481006102 is legit. That's because (probably) musl toolchain on CI is old, could you skip tests on that env?

**Comment by folkertdev** [comment]
ok, rebased again to fix that freebsd problem

> could you skip tests on that env

how would I do this? do you mean a `cfg` for the undefined constants?

**Comment by JohnTitor** [comment]
https://github.com/rust-lang/libc/blob/master/libc-test/build.rs is a place to tweak, like:
https://github.com/rust-lang/libc/blob/68e06adafa8f793b094a23b210d39b27c813600c/libc-test/build.rs#L1739-L1754

**Comment by folkertdev** [comment]
allright, I added logic to ignore the constant for `i686` and `mips`. that should make the tests that failed before pass, but there might be other (cancelled) ci runs that also should skip this constant. Is there a better way to figure that out than trial and error?

**Comment by JohnTitor** [comment]
There's a trade-off, we could disable the fail-fast but we may want to find failures fast.
@bors r+

**Comment by JohnTitor** [comment]
(Another way would be running the same command as CI on your local.)

File: libc-test/build.rs [hunk 1]
```diff
@@ -3724,6 +3724,8 @@ fn test_linux(target: &str) {
                 => true,
             "SCTP_FUTURE_ASSOC" | "SCTP_CURRENT_ASSOC" | "SCTP_ALL_ASSOC" | "SCTP_PEER_ADDR_THLDS_V2" => true, // linux 5.5+
 
+            "HWTSTAMP_TX_ONESTEP_P2P" if i686 || mips || aarch64_musl => true,
```
Comment (JohnTitor) [inline libc-test/build.rs:3727]: I think the below should work like others:

```suggestion
            // FIXME: Requires more recent kernel headers
            "HWTSTAMP_TX_ONESTEP_P2P" if mips || musl => true,
```

Comment (folkertdev) [reply libc-test/build.rs:3728]: yes, fixed. A couple more linux flavors to get through

Comment (JohnTitor) [reply libc-test/build.rs:3728]: The condition doesn't seem to be changed. It should `mips || musl` as we use older kernels for these targets.

Comment (folkertdev) [reply libc-test/build.rs:3728]: though maybe musl is more relevant than aarch here?

Comment (JohnTitor) [reply libc-test/build.rs:3728]: We use the same kernel version for _all_ the musl targets across CI: https://github.com/rust-lang/libc/blob/c64bfc5fb36414e02ad7977b32df765876632383/ci/install-musl.sh#L82
So, CPU arch doesn't involve here, we just need to ignore `musl` on the if cond (the same applies to sparc64).

Comment (folkertdev) [reply libc-test/build.rs:3728]: so `i686 || mips || musl`, or only `musl` even?

Comment (JohnTitor) [reply libc-test/build.rs:3728]: Ah, I mean the condition should be `sparc64 || musl`.

Comment (JohnTitor) [reply libc-test/build.rs:3728]: For sparc64, see:
https://github.com/rust-lang/libc/blob/c64bfc5fb36414e02ad7977b32df765876632383/ci/docker/sparc64-unknown-linux-gnu/Dockerfile#L1-L2

Comment (JohnTitor) [reply libc-test/build.rs:3728]: i686 itself isn't related to the failure, it fails due to the musl toolchain.

Comment (folkertdev) [reply libc-test/build.rs]: makes sense, I've updated it

**Comment by JohnTitor** [comment]
Let's try again, @bors r+

**Comment by folkertdev** [comment]
we're getting closer. I'm not sure what this is trying to do though:

```
RUNNING ALL TESTS
rust[8] = 0 != 8 (C): Rust "hwtstamp_config" -> C
rust[9] = 0 != 9 (C): Rust "hwtstamp_config" -> C
rust[10] = 0 != 10 (C): Rust "hwtstamp_config" -> C
rust[11] = 0 != 11 (C): Rust "hwtstamp_config" -> C
thread 'main' panicked at 'some tests failed', /checkout/target/sparc64-unknown-linux-gnu/debug/build/libc-test-298fd54f1fef40cd/out/main.rs:12:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--test main`

Caused by:
  process didn't exit successfully: `/test-runner-linux sparc64 /checkout/target/sparc64-unknown-linux-gnu/debug/deps/main-973f2bcc439c5f8d` (exit status: 1)
```

should we skip the `hwtstamp_config` struct on this target?

**Comment by folkertdev** [comment]
rather looks like this is an ABI roundtrip test. That this fails is extremely surprising to me, given that the struct is just three `int`s, so nothing tricky is going on. Should we just ignore the roundtrip for `sparc64`?

**Comment by JohnTitor** [comment]
I guess https://github.com/rust-lang/libc/issues/1558 is related. Yeah, skipping would be fine 👍

**Comment by folkertdev** [comment]
allright let's hope that's it

**Pull request merged**